### ISSUE TITLE
Make sure to validate the observation objects.

### ISSUE THREAD

**Issue by jtibshirani** [grf-labs/grf#373]
Currently, we don't validate the user-provided observation objects in R like `Y` and `W`. We should perform at least the following validation:
- Make sure that they are numeric vectors. Currently the observations are allowed to be matrices, which has lead to surprising bugs like #369.
- Ensure that there are no `NA` observations. We don't yet support `NA` values and we suspect that this could be causing the issue seen in #368, where grf can produce `NA` for all predictions.

**Comment by halflearned** [comment]
Thanks, @jtibshirani! 

Related: when the number of rows is zero, the error messages are not very informative. Here's an example.

```
library(grf)

X <- matrix(rnorm(0), nrow = 0, ncol=10)  # Maybe by mistake these matrices were constructed with zero rows
Y <- matrix(rnorm(0), nrow = 0, ncol=1)
W <- matrix(rnorm(0), nrow = 0, ncol=1)

causal_forest(X, Y, W)
```

[This](https://github.com/grf-labs/grf/blob/996f3b3fd4d89761b7b9ed6b0da31dbbba6895ac/core/src/forest/ForestTrainer.cpp#L45) exception will be thrown, triggering the message

```r
  The honesty fraction is too close to 1 or 0, as no observations will be sampled.
```

Sounds like a corner case, but I've encountered this message several times.

**Comment by ras44** [comment]
@jtibshirani I hope this PR is helpful in addressing this issue.  Please let me know if you have any comments!

**Comment by jtibshirani** [comment]
Thanks a lot @ras44, I just had the chance to review the PR.

@halflearned that is confusing, maybe we could open a PR to add a clearer message or file a new issue?

**Comment by jtibshirani** [comment]
Closed by #378.
